app/com/arpnetworking/metrics/portal/scheduling/JobExecutorActor.java

```diff
@@ -18,171 +18,486 @@
+            scheduleTickFor(nextRun.get());
         } else {
-            return TICK_INTERVAL;
+            attemptExecuteAndUpdateRepository(nextRun.get());
         }
     }
 
-    private void scheduleNextTick(final Instant wakeUpBy) {
-        timers().startSingleTimer("TICK", Tick.INSTANCE, timeUntilNextTick(wakeUpBy));
+    private void reload(final Reload<T> message) {
+        final Optional<String> eTag = message.getETag();
+        final CachedJob<T> cachedJob;
+        _periodicMetrics.recordCounter("job_executor_actor_reloads", 1);
+        try {
+            cachedJob = initializeOrEnsureRefMatch(unsafeJobRefCast(message.getJobRef()));
+            if (eTag.isPresent()) {
+                cachedJob.reloadIfOutdated(_injector, eTag.get());
+            } else {
+                cachedJob.reload(_injector);
+            }
+        } catch (final NoSuchElementException error) {
+            LOGGER.warn()
+                    .setMessage("tried to reload job, but job no longer exists in repository")
+                    .addData("ref", message.getJobRef())
+                    .log();
+            killSelf();
+            return;
+        }
+        getSelf().tell(Tick.INSTANCE, getSelf());
     }
 
-    private void executeAndScheduleNextTick(
-            final JobRepository<T> repo,
-            final Organization org,
-            final Job<T> job,
-            final Instant scheduled
-    ) {
-        final Optional<Instant> nextScheduled = job.getSchedule().nextRun(Optional.of(scheduled));
-        repo.jobStarted(job.getId(), org, scheduled);
-        job.execute(getSelf(), scheduled)
-                .handle((result, error) -> {
-                    if (error == null) {
-                        repo.jobSucceeded(job.getId(), org, scheduled, result);
-                    } else {
-                        repo.jobFailed(job.getId(), org, scheduled, error);
-                    }
-                    return null;
-                })
-                .thenApply(whatever -> {
-                    nextScheduled.ifPresent(this::scheduleNextTick);
-                    return null;
-                });
+    private void jobCompleted(final JobCompleted<?> message) {
+        _currentlyExecuting.set(false);
+        if (!_cachedJob.isPresent()) {
+            LOGGER.error()
+                    .setMessage("uninitialized, but got completion message (perhaps from previous life?)")
+                    .addData("scheduled", message.getScheduled())
+                    .addData("outcome", message.getOutcome())
+                    .log();
+            return;
+        }
+        final CachedJob<T> cachedJob = _cachedJob.get();
+        final JobRef<T> ref = cachedJob.getRef();
+
+        @SuppressWarnings("unchecked")
+        final JobCompleted<T> typedMessage = (JobCompleted<T>) message;
+        final JobRepository<T> repo = ref.getRepository(_injector);
+        try {
```

Comment:
This structure and the comments above almost seem to argue for separate message types.